### PR TITLE
ARROW-7717: [CI] Have nightly integration test for Spark's latest release

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

**Comment by kszucs** [comment]
@ursabot crossbow submit test-conda-python-3.6-spark-2.4.4

**Comment by kszucs** [comment]
[AMD64 Conda Crossbow Submit (#88674)](https://ci.ursalabs.org/#builders/98/builds/483) builder has been succeeded.

Revision: 0f4de52129403022aaee438b118132679e19a6a8

Submitted crossbow builds: [ursa-labs/crossbow @ ursabot-468](https://github.com/ursa-labs/crossbow/branches/all?query=ursabot-468)

|Task|Status|
|----|------|
|test-conda-python-3.6-spark-2.4.4|[CircleCI](https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-468-circle-test-conda-python-3.6-spark-2.4.4)|

**Comment by kszucs** [comment]
@ursabot crossbow submit test-conda-python-3.6-spark-2.4.4

**Comment by kszucs** [comment]
[AMD64 Conda Crossbow Submit (#88694)](https://ci.ursalabs.org/#builders/98/builds/484) builder has been succeeded.

Revision: 03a4cfd685c03aaa11924730c08e4f90b0167c14

Submitted crossbow builds: [ursa-labs/crossbow @ ursabot-469](https://github.com/ursa-labs/crossbow/branches/all?query=ursabot-469)

|Task|Status|
|----|------|
|test-conda-python-3.6-spark-2.4.4|[CircleCI](https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-469-circle-test-conda-python-3.6-spark-2.4.4)|

**Comment by kszucs** [comment]
@BryanCutler neither 2.4.4 nor 2.4.5-rc1 spark versions seem compatible with arrow > 0.15. 

Is arrow 0.16 going to be supported only for spark > 3?

**Comment by BryanCutler** [comment]
>@BryanCutler neither 2.4.4 nor 2.4.5-rc1 spark versions seem compatible with arrow > 0.15.
Is arrow 0.16 going to be supported only for spark > 3?

Spark versions 2.4.x need to have the env variable `ARROW_PRE_0_15_IPC_FORMAT=1` to work with pyarrow >= 0.15. See also https://github.com/apache/spark/blob/master/docs/sql-pyspark-pandas-with-arrow.md#compatibility-setting-for-pyarrow--0150-and-spark-23x-24x

**Comment by kszucs** [comment]
Thanks Bryan, I’ll try that then.

**Comment by kszucs** [comment]
@ursabot crossbow submit test-conda-python-3.6-spark-2.4.4 test-conda-python-3.7-spark-master

**Comment by kszucs** [comment]
[AMD64 Conda Crossbow Submit (#90452)](https://ci.ursalabs.org/#builders/98/builds/535) builder has been succeeded.

Revision: ce2060550fdab643808cc8e4dfba4ae052771832

Submitted crossbow builds: [ursa-labs/crossbow @ ursabot-505](https://github.com/ursa-labs/crossbow/branches/all?query=ursabot-505)

|Task|Status|
|----|------|
|test-conda-python-3.6-spark-2.4.4|[CircleCI](https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-505-circle-test-conda-python-3.6-spark-2.4.4)|
|test-conda-python-3.7-spark-master|[CircleCI](https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-505-circle-test-conda-python-3.7-spark-master)|

**Comment by BryanCutler** [comment]
@kszucs I did some local testing and I think there are a number of issues with getting Spark 2.4.4 tests working with the latest pyarrow and pandas >= 1.0.0. It might be better to setup a test for the upcoming release of Spark 3.0.0, the branch has been cut here https://github.com/apache/spark/tree/branch-3.0, that should work fine.

**Comment by kszucs** [comment]
SGTM

**Comment by wesm** [comment]
@BryanCutler @kszucs what do you think about setting up a Spark 3.x integration test?

**Comment by kszucs** [comment]
Definitely, I'm going to roll up this PR and create a jira for spark 3.0 integration testing.

**Comment by kszucs** [comment]
Or just try to add it in this PR.

**Comment by kszucs** [comment]
@github-actions crosbow submit *spark*

**Comment by kszucs** [comment]
@github-actions crossbow submit *spark*

**Comment by kszucs** [comment]
@BryanCutler could you please fix the build errors? I'm not sure what compatibility constraints do we/spark have, perhaps we'll need to stop testing spark 2? The master and spark 3 might error might need to be fixed on spark's side.

**Comment by kszucs** [comment]
@github-actions crossbow submit test-conda-python-3.6-spark-2.4.6

**Comment by BryanCutler** [comment]
Yeah, definitely would be good to test against Spark 3.0. I'll have to work on a patch to handle some of the recent changes to Arrow Java to fix the build.

**Comment by kszucs** [comment]
Thanks You! Although Spark 2 seems to fail as well.

**Comment by BryanCutler** [comment]
Yup, the Spark 2 issues are probably due to the same changes - I'll take a look

**Comment by BryanCutler** [comment]
@kszucs I submitted a patch to fix Java compilation with Spark master and branch-3.0, and tested locally with the latest pyarrow so Spark integration tests should pass for these as of this morning.

For Spark branch-2.4, it will require applying a different patch with the testing script, but maybe we should rethink this strategy a little. Spark 2.4.x will not upgrade Arrow Java anymore, so it would be more correct to take the latest Spark 2.4.6 release out of the box and test that with pyarrow from master. Is that possible to setup, wdyt?

**Comment by kszucs** [comment]
> @kszucs I submitted a patch to fix Java compilation with Spark master and branch-3.0, and tested locally with the latest pyarrow so Spark integration tests should pass for these as of this morning.

Thanks Bryan!

> 
> For Spark branch-2.4, it will require applying a different patch with the testing script, but maybe we should rethink this strategy a little. Spark 2.4.x will not upgrade Arrow Java anymore, so it would be more correct to take the latest Spark 2.4.6 release out of the box and test that with pyarrow from master. Is that possible to setup, wdyt?

I'm not sure that I understand you correctly :)
- keep arrow master <=> spark branch-3.0
- keep arrow master <=> spark master
- and keep arrow master <=> spark 2.4.6 but execute only the pyarrow tests?

**Comment by BryanCutler** [comment]
I mean the current process for integration tests with the master branch is to build Spark with Arrow Java master, then run Java and Python tests. That process is good for Spark master and branch-3.0, but with Spark branch-2.4, it doesn't make too much sense to rebuild Spark with Arrow Java master. This is because the dependency on Arrow Java is pretty much frozen for branch-2.4 (unless there is a serious enough bug), but different versions of pyarrow can be used. It would be best to take the latest Spark 2.4.6 release out of the box and run the pyspark tests from the script with pyarrow from master. We are not really concerned about Java compatibility since it can't be upgraded anyway, but that still allows us to flush out any Python problems.

**Comment by BryanCutler** [comment]
@kszucs the nightly against Spark master have been passing. Do you think you could update this to just add the test against branch-3.0 and remove branch-2.4 for now? I'm not sure if the bot will pick it up from this PR, but I'll try to kick it off a test run.

File: dev/tasks/tasks.yml [hunk 1]
```diff
@@ -1833,12 +1833,32 @@ tasks:
...
     template: docker-tests/github.linux.yml
     params:
       env:
-        PYTHON: 3.7
+        PYTHON: 3.6
+        SPARK: "v2.4.6"
+      # use the master branch of spark, so prevent reusing any layers
+      run: --no-leaf-cache --env ARROW_PRE_0_15_IPC_FORMAT=1 conda-python-spark
+
+  test-conda-python-3.7-spark-branch-3.0:
```
Comment (BryanCutler) [inline dev/tasks/tasks.yml:1846]: looks like this should be named python-3.8 also

**Comment by BryanCutler** [comment]
@github-actions crossbow submit test-conda-python-3.7-spark-branch-3.0

**Review by BryanCutler — COMMENTED** [review]
LGTM, I think we could skip branch-2.4 for now and branch-3.0 should be working

File: dev/tasks/tasks.yml (hunk 1 above)
Comment (kszucs) [reply dev/tasks/tasks.yml:1846]: Fixing it.

**Comment by kszucs** [comment]
@github-actions crossbow submit  test-*spark*

**Review by kszucs — COMMENTED** [review]
+1, thanks @BryanCutler!

**Comment by BryanCutler** [comment]
This is great, thanks @kszucs !

**Pull request closed without merging**